libc/string_lenchr.c

```diff
@@ -0,0 +1,143 @@
+ * Phoenix-RTOS
+ *
+ * POSIX.1-2017 standard library functions tests
+ * HEADER:
+ * 	string.h
+ * TESTED:
+ * 	strlen()
+ * 	strnlen()
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Damian Modzelewski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <limits.h>
+#include <unity_fixture.h>
+
+
+TEST_GROUP(string_len);
+
+
+TEST_SETUP(string_len)
+{
+}
+
+
+TEST_TEAR_DOWN(string_len)
+{
+}
+
+
+TEST(string_len, ascii)
+{
+	const char empty[] = "";
+	const char pangram[] = "The quick brown fox jumps over the lazy dog";
+	const char torn[] = "foo\0bar";
+	const char doubleNul[] = "\0\0abc";
+	const char specials[] = "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~";
+	const char whites[] = " \v\t\r\n";
+	char asciiSet[128] = { 0 };
+	int sz, i;
+
+	TEST_ASSERT_EQUAL_INT(0, strlen(""));
+
+	/* pangram with a whole alphabet set */
+	sz = sizeof(pangram) - 1;
+	TEST_ASSERT_EQUAL_INT(sz, strlen(pangram));
+	TEST_ASSERT_EQUAL_INT(sz - 1, strnlen(pangram, sz - 1));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(pangram, sz));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(pangram, sz + 1));
+
+	/* text with null character in the middle*/
+	sz = (sizeof(torn) / 2) - 1;
+	TEST_ASSERT_EQUAL_INT(sz, strlen(torn));
+	TEST_ASSERT_EQUAL_INT(sz - 1, strnlen(torn, 3 - 1));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(torn, 3));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(torn, 3 + 1));
+
+	/* end of string */
+	sz = 0;
+	TEST_ASSERT_EQUAL_INT(sz, strlen(doubleNul));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(doubleNul, 0));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(doubleNul, 1));
+	sz = 0;
+	TEST_ASSERT_EQUAL_INT(sz, strlen(doubleNul));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(empty, 0));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(empty, 1));
+
+	/* special characters */
+	sz = sizeof(specials) - 1;
+	TEST_ASSERT_EQUAL_INT(sz, strlen(specials));
+	TEST_ASSERT_EQUAL_INT(sz - 1, strnlen(specials, sz - 1));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(specials, sz));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(specials, sz + 1));
+
+	/* white spaces */
+	sz = sizeof(whites) - 1;
+	TEST_ASSERT_EQUAL_INT(sz, strlen(whites));
+	TEST_ASSERT_EQUAL_INT(sz - 1, strnlen(whites, sz - 1));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(whites, sz));
+	TEST_ASSERT_EQUAL_INT(sz, strnlen(whites, sz + 1));
+
+	/*Checking ascii charset*/
+	for (i = 1; i < sizeof(asciiSet) - 1; i++) {
+		asciiSet[i - 1] = i;
+		printf("\n%s\n", asciiSet);
+	}
+
+	TEST_ASSERT_EQUAL_INT(sizeof(asciiSet) - 2, strlen(asciiSet));
+}
+
+
+TEST(string_len, not_ascii)
+{
+	const char notAsciiString[] = "♦♥♣♠◊⊗こんにちは❉❉⌨⌨⌨⌨⌨⌨⌨⌨❉❉";
+	char notAsciiSet[128];
+	int i;
+
+	/* Checking ability to read the chars out of ascii charset*/
+	TEST_ASSERT_EQUAL_INT(sizeof(notAsciiString) - 1, strlen(notAsciiString));
+
+	/* Checking something similar to ascii set */
+	for (i = 1; i < sizeof(notAsciiSet); i++) {
+		notAsciiSet[i - 1] = i * -1;
```

Comment:
The previous approach worked, but let's do it this way. I think it will be more readable.

```suggestion
	unsigned char notAsciiSet[129];
	int i;

	/* Checking ability to read the chars out of ascii charset */
	TEST_ASSERT_EQUAL_INT(sizeof(notAsciiString) - 1, strlen(notAsciiString));

	/* Checking out of ASCII bytes */
	for (i = 128; i <= 255; i++) {
		notAsciiSet[i - 128] = i;
```